### PR TITLE
Refactor results into common CompilationResult hierarchy

### LINKED ISSUE #129: All compilers should probably return a common Result type
It's hard to abstract assertions over multiplatform compilations without it.

### PR DESCRIPTION
**Pull request by ZacSweers**
This creates a new sealed `CompilationResult` hierarchy that includes subtypes for JS and JVM results. This should be a mostly source-compatible (but not binary compatible) change.

One important source compatibility change is that now JS files output is just called `jsFiles` rather than the borrowed JVM "classesAndResources" name from before.

Resolves #129

### CODE REVIEW

**Comment by ZacSweers** [comment]
@jakewharton would be curious for your thoughts on this!

File: core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package com.facebook.buck.jvm.java.javax.com.tschuchort.compiletesting
```
Comment (JakeWharton) [inline core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt:1]: Uhhhhh

File: core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import com.tschuchort.compiletesting.KotlinCompilation
+import com.tschuchort.compiletesting.KotlinJsCompilation
+import com.tschuchort.compiletesting.listFilesRecursively
+import java.io.File
+import java.net.URLClassLoader
+import org.jetbrains.kotlin.compiler.plugin.ExperimentalCompilerApi
+
+/** Result of the compilation. */
+@ExperimentalCompilerApi
+sealed class CompilationResult(
```
Comment (JakeWharton) [inline core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt:12]: Could be `interface`. Don't think you gain anything by using a class here except maybe the future pain of the rigidity of inheritance.

**Review by JakeWharton — APPROVED** [review]
Nice!

File: core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt (hunk 1 above)
Comment (ZacSweers) [reply core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt:1]: oh my

Comment (ZacSweers) [reply core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt]: [aa38fac](https://github.com/ZacSweers/kotlin-compile-testing/pull/136/commits/aa38fac0fc55a88799f1a6698649b61402d08bbc)

File: core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt (hunk 2 above)
Comment (ZacSweers) [reply core/src/main/kotlin/com/tschuchort/compiletesting/CompilationResult.kt]: [54c624c](https://github.com/ZacSweers/kotlin-compile-testing/pull/136/commits/54c624c9503f431295730894e3f13a955e4398e2)

**Pull request merged**